File: backend/monitor/views/observations.py
"""观测记录查询、修正合计明细、区间补算与人工重放控制接口。"""

from decimal import Decimal

from django.db.models import Q
from django.shortcuts import get_object_or_404

from .base import AdminAPIView, PageAccessAPIView, error, ok
from ..api_auth import APIKeyAuthentication
from ..access import visible_accounts_for, visible_participant_ids
from ..billing_correction.domain import BillingCorrectionRules
from ..billing_correction.observations import interval_corrections
from ..billing_correction.rules import CORRECTION_SETTINGS, corrections_digest
from ..fast_correction.repair import calculate_missing_fast_correction
from ..integrations.sub2api import Sub2APIError
from ..reporting import iso, snapshot_data
from .query_params import monitored_account_query
from .record_helpers import paginated_rows, query_datetime
from ..models import (
    AppSettings,
    Observation,
    MonitoredAccount,
    PagePermission,
    Participant,
    Sub2APIUserUsageSample,
)
from ..replay import (
    clear_manual_start,
    exclude_observation,
    rebuild_current_interval,
    restore_observation,
    set_manual_start,
)


class ObservationListView(PageAccessAPIView):
    required_page_permissions = (PagePermission.OBSERVATIONS,)

    def get(self, request):
        config = AppSettings.load()
        visible_ids = visible_participant_ids(request.user)
        try:
            account = monitored_account_query(request)
        except ValueError as exc:
            return error(str(exc), 400)
        queryset = Observation.objects.select_related(
            "manual_start_end", "billing_capture"
        ).prefetch_related("participant_snapshots__participant", "fast_corrections", "billing_capture__facts")
        if account is not None:
            queryset = queryset.filter(account_id=account.fact_key)
        elif not request.user.is_staff:
            queryset = queryset.none()
        try:
            observed_from = query_datetime(request, "from")
            observed_to = query_datetime(request, "to")
        except ValueError as exc:
            return error(str(exc), 400)
        if observed_from:
            queryset = queryset.filter(observed_at__gte=observed_from)
        if observed_to:
            queryset = queryset.filter(observed_at__lte=observed_to)

        source = request.query_params.get("source")
        valid_sources = {value for value, _label in Observation.SOURCE_CHOICES}
        if source:
            if source not in valid_sources:
                return error("来源筛选值无效", 400)
            queryset = queryset.filter(source=source)

        query_mode = request.query_params.get("query_mode")
        if query_mode:
            if query_mode not in {"passive", "direct"}:
                return error("查询方式筛选值无效", 400)
            if query_mode == "passive":
                queryset = queryset.filter(raw_window__query_mode="passive")
            else:
                # 旧观测尚未写 query_mode 时，展示层一直将它解释为上游直查。
                queryset = queryset.filter(
                    Q(raw_window__query_mode="direct")
                    | ~Q(raw_window__has_key="query_mode")
                )

        total = queryset.count()
        excluded_count = queryset.filter(excluded_at__isnull=False).count()
        valid_count = queryset.filter(valid_sample=True).count()
        passive_count = queryset.filter(raw_window__query_mode="passive").count()
        rows, pagination = paginated_rows(request, queryset)
        result = []
        rules = BillingCorrectionRules(config)
        for item in rows:
            correction = interval_corrections(item, config, rules=rules)
            result.append(
                {
                    "id": item.id,
                    "observed_at": iso(item.observed_at),
                    "source": item.source,
                    "provider": item.raw_window.get("provider", "sub2api"),
                    "account_id": item.account_id,
                    "attribution_started_at": iso(item.attribution_started_at),
                    "upstream_resets_at": iso(item.upstream_resets_at),
                    "upstream_used_percent": float(item.upstream_used_percent),
                    "interval_used_percent": float(item.interval_used_percent),
                    "raw_selected_total_cost": float(
                        item.raw_selected_total_cost
                    ),
                    "selected_total_cost": float(item.selected_total_cost),
                    "cost_window_started_at": iso(
                        item.cost_window_started_at
                    ),
                    "cost_window_ended_at": iso(item.cost_window_ended_at),
                    "interval_cost_started_at": iso(
                        item.interval_cost_started_at
                    ),
                    "interval_cost": (
                        float(item.interval_cost(config.cost_basis))
                        if item.interval_cost(config.cost_basis) is not None
                        else None
                    ),
                    "interval_cost_source": item.interval_cost_source,
                    "normalized_total_cost": float(
                        item.normalized_cost(config.cost_basis)
                    ),
                    "delta_percent": (
                        float(item.delta_percent)
                        if item.delta_percent is not None
                        else None
                    ),
                    "delta_cost": (
                        float(item.delta_cost)
                        if item.delta_cost is not None
                        else None
                    ),
                    "sample_usd_per_percent": (
                        float(item.sample_usd_per_percent)
                        if item.sample_usd_per_percent is not None
                        else None
                    ),
                    "effective_usd_per_percent": float(
                        item.effective_usd_per_percent
                    ),
                    "estimated_used_percent": float(
                        item.estimated_used_percent
                    ),
                    "capacity_lower_usd": (
                        float(item.capacity_lower_usd)
                        if item.capacity_lower_usd is not None
                        else None
                    ),
                    "capacity_upper_usd": (
                        float(item.capacity_upper_usd)
                        if item.capacity_upper_usd is not None
                        else None
                    ),
                    "model_diagnostics": item.model_diagnostics,
                    **correction.payload(),
                    "correction_total_usd": float(correction.amounts.total) if correction.calculated else None,
                    "fast_correction_usd": float(correction.amounts.fast) if correction.calculated or correction.legacy_fast_only else None,
                    # Retain the legacy field's FAST-only meaning. The total
                    # has its own completeness flag and can still need backfill.
                    "fast_correction_calculated": correction.calculated or correction.legacy_fast_only,
                    "valid_sample": item.valid_sample,
                    "sample_note": item.sample_note,
                    "rate_method": item.raw_window.get(
                        "rate_method",
                        "incremental_legacy",
                    ),
                    "query_mode": item.raw_window.get("query_mode", "direct"),
                    "snapshot_sampled_at": item.raw_window.get("sampled_at"),
                    "excluded": item.excluded_at is not None,
                    "excluded_at": iso(item.excluded_at),
                    "exclusion_reason": item.exclusion_reason,
                    "exclusion_source": item.exclusion_source,
                    "is_manual_start": item.is_manual_start,
                    "manual_start_reason": item.manual_start_reason,
                    "manual_start_set_at": iso(item.manual_start_set_at),
                    "manual_start_end_id": item.manual_start_end_id,
                    "manual_start_end_observed_at": iso(
                        item.manual_start_end.observed_at
                        if item.manual_start_end is not None
                        else None
                    ),
                    "participants": [
                        snapshot_data(snapshot)
                        for snapshot in item.participant_snapshots.all()
                        if visible_ids is None
                        or snapshot.participant_id in visible_ids
                    ],
                }
            )
        return ok(
            {
                "account": (
                    {
                        "id": account.id,
                        "provider": account.provider,
                        "source_account_id": account.source_account_id,
                        "external_account_id": account.external_account_id,
                        "name": account.name,
                    }
                    if account is not None
                    else None
                ),
                "items": result,
                "corrections_available": account is None or account.provider == "sub2api",
                "fast_correction_enabled": bool(
                    account is None
                    or (
                        account.provider == "sub2api"
                        and config.fast_correction_enabled
                    )
                ),
                "pagination": pagination,
                "summary": {
                    "total": total,
                    "valid_count": valid_count,
                    "passive_count": passive_count,
                    "excluded_count": excluded_count,
                },
            }
        )


class ReadOnlyObservationListView(ObservationListView):
    """External API-key view exposing paginated observation records."""

    authentication_classes = [APIKeyAuthentication]
    http_method_names = ["get", "head", "options"]


class ObservationFastCorrectionDetailView(PageAccessAPIView):
    """展示当前规则从原始事实计算的三项修正；旧 FAST-only 记录明确标识。"""

    required_page_permissions = (PagePermission.OBSERVATIONS,)

    def get(self, request, observation_id: int):
        config = AppSettings.load()
        observations = Observation.objects.select_related("billing_capture").prefetch_related("fast_corrections", "billing_capture__facts")
        if not request.user.is_staff:
            observations = observations.filter(account_id__in=[item.fact_key for item in visible_accounts_for(request.user)])
        observation = get_object_or_404(observations, pk=observation_id)
        source_account = MonitoredAccount.for_fact_key(observation.account_id)
        if observation.account_id < 0 or (
            source_account is not None
            and source_account.provider in {"cpa", "gpt_load"}
        ):
            label = (
                "GPT-Load"
                if source_account is not None
                and source_account.provider == "gpt_load"
                else "CPA"
            )
            return error(f"{label} 观测不使用 Sub2API 修正明细", 400)
        correction = interval_corrections(observation, config, include_models=True)
        visible_ids = visible_participant_ids(request.user)
        participant_queryset = Participant.objects.filter(sub2api_user_id__in=correction.users)
        if visible_ids is not None:
            participant_queryset = participant_queryset.filter(id__in=visible_ids)
        participants = {item.sub2api_user_id: item for item in participant_queryset}
        rows = [row for row in correction.users.values() if visible_ids is None or row.user_id in participants]
        samples = {item.sub2api_user_id: item for item in Sub2APIUserUsageSample.objects.filter(
            account_id=observation.account_id, observed_at=observation.observed_at,
            sub2api_user_id__in=[row.user_id for row in rows],
        )}
        users = []
        for row in sorted(rows, key=lambda row: row.user_id):
            sample, participant = samples.get(row.user_id), participants.get(row.user_id)
            username, email = (sample.username, sample.email) if sample else ("", "")
            users.append({
                "sub2api_user_id": row.user_id, "username": username, "email": email,
                "display_name": username or email or (participant.name if participant else "") or f"用户 {row.user_id}",
                "request_count": row.request_count, "fast_request_count": row.fast_request_count,
                "non_fast_request_count": max(0, row.request_count - row.fast_request_count) if row.request_count is not None else None,
                "fast_billed_cost_usd": float(row.fast_raw_cost),
                "correction_usd": float(row.amounts.total),
                "corrected_fast_cost_usd": float(row.fast_raw_cost + row.amounts.fast),
                "raw_cost_usd": float(row.raw_cost) if correction.facts_complete else None,
                "corrected_cost_usd": float(row.raw_cost + row.amounts.total) if correction.facts_complete else None,
                "unknown_long_context_request_count": row.unknown_long_context_request_count,
                **row.amounts.payload(),
            })
        request_count = sum(row.request_count for row in rows) if all(row.request_count is not None for row in rows) else None
        if not rows and correction.request_count is None:
            request_count = None
        fast_count = sum(row.fast_request_count for row in rows)
        fast_raw_cost = sum((row.fast_raw_cost for row in rows), Decimal("0"))
        capture = getattr(observation, "billing_capture", None)
        return ok({
            "observation_id": observation.id,
            "started_at": iso(capture.started_at if capture else observation.fast_correction_started_at),
            "ended_at": iso(observation.observed_at), "calculated": correction.calculated,
            "cost_basis": config.cost_basis,
            "cost_basis_label": "实际扣费" if config.cost_basis == "actual" else "标准扣费",
            "request_count": request_count, "fast_request_count": fast_count,
            "non_fast_request_count": max(0, request_count - fast_count) if request_count is not None else None,
            "fast_billed_cost_usd": float(fast_raw_cost),
            "correction_usd": float(correction.amounts.total),
            "corrected_fast_cost_usd": float(fast_raw_cost + correction.amounts.fast),
            "raw_cost_usd": float(correction.raw_cost) if correction.raw_cost is not None else None,
            "corrected_cost_usd": float(correction.raw_cost + correction.amounts.total) if correction.raw_cost is not None else None,
            "collection_error": str(observation.raw_window.get("fast_correction_error") or "") if not correction.facts_complete else "",
            "users": users, "model_details": correction.model_details,
            "calculation_order": ["fast", "long_context", "model"],
            "rules_digest": corrections_digest(config),
            "rules": {name: getattr(config, name) for name in sorted(CORRECTION_SETTINGS)},
            **correction.payload(),
        })


class ReadOnlyObservationFastCorrectionDetailView(
    ObservationFastCorrectionDetailView
):
    """External API-key view exposing persisted FAST correction facts."""

    authentication_classes = [APIKeyAuthentication]
    http_method_names = ["get", "head", "options"]

class ObservationFastCorrectionCalculateView(AdminAPIView):
    """从上游补齐单条缺失/旧 FAST-only 区间的原始请求并重算三项修正。"""

    def post(self, _request, observation_id: int):
        config = AppSettings.load()
        observation = get_object_or_404(Observation, pk=observation_id)
        account = MonitoredAccount.for_fact_key(observation.account_id)
        if account is not None and account.provider in {"cpa", "gpt_load"}:
            label = "GPT-Load" if account.provider == "gpt_load" else "CPA"
            return error(f"{label} 请求成本已在采集时计价", 400)
        try:
            result = calculate_missing_fast_correction(observation, config)
        except ValueError as exc:
            return error(str(exc), 400)
        except Sub2APIError as exc:
            return error(str(exc), 502)
        return ok(result)


class ObservationRebuildView(AdminAPIView):
    """保留原始采样与人工标记，仅重建当前区间的全部派生结论。"""

    def post(self, request):
        config = AppSettings.load()
        try:
            account = monitored_account_query(request, enabled_only=True)
        except ValueError as exc:
            return error(str(exc), 400)
        if account is None:
            return error("尚未配置启用的监控账号", 400)
        replay, replay_from = rebuild_current_interval(
            account.fact_key,
            config,
        )
        return ok(
            {
                **replay.as_dict(),
                "replay_started_at": iso(replay_from),
            }
        )


class ObservationExclusionView(AdminAPIView):
    """排除一条校准记录，并从最早受影响区间向后重放。"""

    def post(self, request, observation_id: int):
        observation = get_object_or_404(Observation, pk=observation_id)
        reason = request.data.get("reason", "管理员手动排除")
        if not isinstance(reason, str):
            return error("排除原因格式无效", 400)
        result = exclude_observation(observation, reason)
        return ok(result)


class ObservationRestoreView(AdminAPIView):
    """恢复一条排除记录，并从最早受影响区间向后重放。"""

    def post(self, _request, observation_id: int):
        observation = get_object_or_404(Observation, pk=observation_id)
        return ok(restore_observation(observation))


class ObservationManualStartView(AdminAPIView):
    """设置或取消最高优先级的管理员观测起点区间。"""

    def post(self, request, observation_id: int):
        observation = get_object_or_404(Observation, pk=observation_id)
        reason = request.data.get("reason", "")
        if not isinstance(reason, str):
            return error("起点说明格式无效", 400)
        end_observation_id = request.data.get(
            "end_observation_id",
            observation.id,
        )
        if isinstance(end_observation_id, bool) or not isinstance(
            end_observation_id,
            int,
        ):
            return error("起点区间终点记录无效", 400)
        end_observation = get_object_or_404(
            Observation,
            pk=end_observation_id,
        )
        try:
            result = set_manual_start(
                observation,
                reason,
                end_observation=end_observation,
            )
        except ValueError as exc:
            return error(str(exc), 400)
        return ok(result)

    def delete(self, _request, observation_id: int):
        observation = get_object_or_404(Observation, pk=observation_id)
        return ok(clear_manual_start(observation))

File: frontend/src/demo/handlers/observations.ts
import { correctionCalculated } from "@/utils/corrections";
import type {
  FastCorrectionCalculateResult,
  FastCorrectionDetail,
  Observation,
  ObservationListData,
  ObservationRebuildResult,
} from "@/types/observations";

import type { DemoRequestContext } from "../backend";
import { demoIdentity, saveDemoState, type DemoState } from "../state";

function observationsData(context: DemoRequestContext): ObservationListData {
  const { state, url, paginate } = context;
  const accountId = Number(url.searchParams.get("account_id"));
  const account =
    state.monitoredAccounts.find((item) => item.id === accountId) ??
    state.monitoredAccounts.find((item) => item.enabled) ??
    state.monitoredAccounts[0];
  let items = [...state.observations].reverse().map((item) => ({
    ...item,
    correction_calculated: item.correction_calculated ?? false,
    legacy_fast_only: item.legacy_fast_only ?? item.fast_correction_calculated,
    account_id: account?.external_account_id ?? item.account_id,
  }));
  if (account?.provider === "cpa" || account?.provider === "gpt_load") {
    items = [];
  }
  const from = url.searchParams.get("from");
  const to = url.searchParams.get("to");
  const source = url.searchParams.get("source");
  const queryMode = url.searchParams.get("query_mode");
  if (from) items = items.filter((item) => item.observed_at >= from);
  if (to) items = items.filter((item) => item.observed_at <= to);
  if (source) items = items.filter((item) => item.source === source);
  if (queryMode) items = items.filter((item) => item.query_mode === queryMode);
  const page = paginate(items);
  return {
    ...page,
    account: account
      ? {
          id: account.id,
          provider: account.provider,
          source_account_id: account.source_account_id,
          external_account_id: account.external_account_id,
          name: account.name,
        }
      : null,
    corrections_available: account?.provider === "sub2api",
    summary: {
      total: items.length,
      valid_count: items.filter((item) => item.valid_sample).length,
      passive_count: items.filter((item) => item.query_mode === "passive")
        .length,
      excluded_count: items.filter((item) => item.excluded).length,
    },
  };
}

function fastCorrectionData(
  state: DemoState,
  observation: Observation,
): FastCorrectionDetail {
  if (observation.correction_source !== "local") {
    return {
      observation_id: observation.id,
      correction_source: observation.correction_source,
      pricing_epoch: observation.pricing_epoch,
      message:
        observation.correction_source === "upstream"
          ? "上游 Sub2API 已修正"
          : "此记录不进行本地修正。",
    };
  }
  const totalRequests = Math.max(
    12,
    Math.round((observation.delta_cost ?? 8) * 9),
  );
  const fastRequests = Math.round(totalRequests * 0.34);
  const fastCost = observation.fast_correction_usd ?? 0;
  return {
    observation_id: observation.id,
    correction_source: "local",
    pricing_epoch: observation.pricing_epoch,
    started_at: observation.interval_cost_started_at,
    ended_at: observation.observed_at,
    calculated: observation.correction_calculated ?? false,
    cost_basis: "actual",
    cost_basis_label: "实际扣费",
    request_count: totalRequests,
    fast_request_count: fastRequests,
    non_fast_request_count: totalRequests - fastRequests,
    fast_billed_cost_usd: fastCost * 2,
    correction_usd: fastCost,
    fast_correction_usd: fastCost,
    correction_total_usd: fastCost,
    long_context_correction_usd: 0,
    model_correction_usd: 0,
    correction_calculated: observation.correction_calculated ?? false,
    correction_facts_complete: observation.correction_facts_complete ?? false,
    legacy_fast_only:
      observation.legacy_fast_only ?? observation.fast_correction_calculated,
    corrected_fast_cost_usd: fastCost * 3,
    collection_error: "",
    users: state.participants
      .filter((p) => p.sub2api_user_id != null)
      .map((participant, index) => {
        const requestCount = Math.round(
          totalRequests * [0.39, 0.35, 0.26][index],
        );
        const participantFast = Math.round(requestCount * 0.34);
        const billed = fastCost * [0.39, 0.35, 0.26][index] * 2;
        return {
          sub2api_user_id: participant.sub2api_user_id!,
          username: participant.sub2api_username,
          email: participant.sub2api_email,
          display_name: participant.name,
          request_count: requestCount,
          fast_request_count: participantFast,
          non_fast_request_count: requestCount - participantFast,
          fast_billed_cost_usd: billed,
          correction_usd: billed * 0.5,
          fast_correction_usd: billed * 0.5,
          correction_total_usd: billed * 0.5,
          long_context_correction_usd: 0,
          model_correction_usd: 0,
          corrected_fast_cost_usd: billed * 1.5,
        };
      }),
  };
}

export function handleObservations(
  context: DemoRequestContext,
): Response | null {
  const { method, pathname, payload, state, ok, fail } = context;
  if (method === "GET" && pathname === "observations") {
    return ok(observationsData(context));
  }
  const fastCalculateMatch =
    /^observations\/(\d+)\/fast-correction\/calculate$/.exec(pathname);
  if (method === "POST" && fastCalculateMatch) {
    if (!demoIdentity()?.is_staff) return fail("没有管理员权限", 403);
    const observation = state.observations.find(
      (item) => item.id === Number(fastCalculateMatch[1]),
    );
    if (!observation) return fail("观测不存在", 404);
    // Demo captures only simulate status; they never query Sub2API.
    if (
      !correctionCalculated({
        ...observation,
        correction_calculated: observation.correction_calculated ?? false,
      })
    ) {
      observation.fast_correction_usd ??= Number(
        (observation.selected_total_cost * 0.036).toFixed(6),
      );
      observation.fast_correction_calculated = true;
      observation.correction_calculated = true;
      observation.correction_facts_complete = true;
      observation.legacy_fast_only = false;
      observation.correction_total_usd = observation.fast_correction_usd;
      observation.long_context_correction_usd = 0;
      observation.model_correction_usd = 0;
      saveDemoState(state);
    }
    return ok({
      observation_id: observation.id,
      fast_correction_usd: observation.fast_correction_usd ?? 0,
      fast_correction_calculated: true,
      correction_calculated: true,
      correction_facts_complete: true,
      legacy_fast_only: false,
      correction_total_usd: observation.correction_total_usd,
      long_context_correction_usd: observation.long_context_correction_usd,
      model_correction_usd: observation.model_correction_usd,
    } satisfies FastCorrectionCalculateResult);
  }
  const fastMatch = /^observations\/(\d+)\/fast-correction$/.exec(pathname);
  if (method === "GET" && fastMatch) {
    const observation = state.observations.find(
      (item) => item.id === Number(fastMatch[1]),
    );
    return observation
      ? ok(fastCorrectionData(state, observation))
      : fail("观测不存在", 404);
  }
  const exclusionMatch = /^observations\/(\d+)\/(exclude|restore)$/.exec(
    pathname,
  );
  if (method === "POST" && exclusionMatch) {
    const observation = state.observations.find(
      (item) => item.id === Number(exclusionMatch[1]),
    );
    if (!observation) return fail("观测不存在", 404);
    const excluded = exclusionMatch[2] === "exclude";
    observation.excluded = excluded;
    observation.valid_sample = !excluded;
    observation.excluded_at = excluded ? state.clock : null;
    observation.exclusion_reason = excluded
      ? String(payload.reason ?? "管理员手动排除")
      : "";
    observation.exclusion_source = excluded ? "manual" : "";
    saveDemoState(state);
    return ok({ replayed: true });
  }
  const manualStartMatch = /^observations\/(\d+)\/manual-start$/.exec(pathname);
  if (manualStartMatch && (method === "POST" || method === "DELETE")) {
    const observation = state.observations.find(
      (item) => item.id === Number(manualStartMatch[1]),
    );
    if (!observation) return fail("观测不存在", 404);
    if (method === "POST") {
      const endObservationId =
        typeof payload.end_observation_id === "number"
          ? payload.end_observation_id
          : observation.id;
      const endObservation = state.observations.find(
        (item) => item.id === endObservationId,
      );
      if (!endObservation) return fail("起点区间终点记录不存在", 404);
      const compare = (left: Observation, right: Observation) =>
        Date.parse(left.observed_at) - Date.parse(right.observed_at) ||
        left.id - right.id;
      if (
        observation.account_id !== endObservation.account_id ||
        compare(endObservation, observation) < 0
      ) {
        return fail("起点区间终点记录无效", 400);
      }
      for (const item of state.observations) {
        if (item.id === observation.id || !item.is_manual_start) continue;
        const existingEnd =
          state.observations.find(
            (candidate) => candidate.id === item.manual_start_end_id,
          ) ?? item;
        const overlaps =
          compare(item, endObservation) <= 0 &&
          compare(observation, existingEnd) <= 0;
        if (!overlaps) continue;
        const contained =
          compare(observation, item) <= 0 &&
          compare(existingEnd, endObservation) <= 0;
        if (!contained) return fail("起点区间与现有区间部分重叠", 400);
        item.is_manual_start = false;
        item.manual_start_reason = "";
        item.manual_start_set_at = null;
        item.manual_start_end_id = null;
        item.manual_start_end_observed_at = null;
      }
      observation.is_manual_start = true;
      observation.manual_start_reason = String(payload.reason ?? "");
      observation.manual_start_set_at = state.clock;
      observation.manual_start_end_id = endObservation.id;
      observation.manual_start_end_observed_at = endObservation.observed_at;
    } else {
      observation.is_manual_start = false;
      observation.manual_start_reason = "";
      observation.manual_start_set_at = null;
      observation.manual_start_end_id = null;
      observation.manual_start_end_observed_at = null;
    }
    saveDemoState(state);
    return ok({ replayed: true });
  }
  if (method === "POST" && pathname === "observations/rebuild") {
    const result: ObservationRebuildResult = {
      rebuilt_observations: state.observations.length,
      automatic_exclusions: state.observations.filter(
        (item) => item.exclusion_source === "automatic",
      ).length,
      inferred_intervals: 0,
      latest_observation_id: state.observations.at(-1)?.id ?? null,
      replay_started_at: state.periods[0].startedAt,
    };
    return ok(result);
  }
  return null;
}
